File: modal-go/auth_token_manager.go
package modal

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	pb "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
)

const (
	// Start refreshing this many seconds before the token expires
	RefreshWindow = 5 * 60
	// If the token doesn't have an expiry field, default to current time plus this value (not expected).
	DefaultExpiryOffset = 20 * 60
)

type TokenAndExpiry struct {
	token  string
	expiry int64
}

// AuthTokenManager manages authentication tokens, refreshing them lazily
// when GetToken is called. Tokens are refreshed when expired or within
// RefreshWindow seconds of expiry.
type AuthTokenManager struct {
	client pb.ModalClientClient
	logger *slog.Logger

	tokenAndExpiry atomic.Value
	refreshMu      sync.Mutex
}

func NewAuthTokenManager(client pb.ModalClientClient, logger *slog.Logger) *AuthTokenManager {
	manager := &AuthTokenManager{
		client: client,
		logger: logger,
	}

	manager.tokenAndExpiry.Store(TokenAndExpiry{
		token:  "",
		expiry: 0,
	})

	return manager
}

// GetToken returns a valid auth token, fetching or refreshing as needed.
//
// Three states:
//  1. Valid token (not near expiry): returned immediately, no locking.
//  2. No token or expired: all callers block until a fresh token is fetched.
//     Only one goroutine makes the RPC; others wait on the mutex then see the
//     new token via a double-check.
//  3. Valid but within RefreshWindow of expiry: one goroutine refreshes
//     (blocking only itself); concurrent callers get the old, still-valid token.
func (m *AuthTokenManager) GetToken(ctx context.Context) (string, error) {
	data := m.tokenAndExpiry.Load().(TokenAndExpiry)

	if data.token == "" || isExpired(data) {
		if err := m.lockedRefresh(ctx); err != nil {
			return "", err
		}
		return m.tokenAndExpiry.Load().(TokenAndExpiry).token, nil
	}

	if needsRefresh(data) {
		if m.refreshMu.TryLock() {
			data = m.tokenAndExpiry.Load().(TokenAndExpiry)
			if needsRefresh(data) {
				if _, err := m.FetchToken(ctx); err != nil {
					m.logger.ErrorContext(ctx, "refreshing auth token", "error", err)
				}
			}
			m.refreshMu.Unlock()
		}
	}

	return m.tokenAndExpiry.Load().(TokenAndExpiry).token, nil
}

// lockedRefresh blocks until the mutex is acquired, then refreshes if still needed.
func (m *AuthTokenManager) lockedRefresh(ctx context.Context) error {
	m.refreshMu.Lock()
	defer m.refreshMu.Unlock()

	data := m.tokenAndExpiry.Load().(TokenAndExpiry)
	if data.token != "" && !needsRefresh(data) {
		return nil
	}
	_, err := m.FetchToken(ctx)
	return err
}

// FetchToken fetches a new token using AuthTokenGet() and stores it.
func (m *AuthTokenManager) FetchToken(ctx context.Context) (string, error) {
	resp, err := m.client.AuthTokenGet(ctx, &pb.AuthTokenGetRequest{})
	if err != nil {
		return "", fmt.Errorf("AuthTokenGet: %w", err)
	}

	token := resp.GetToken()
	if token == "" {
		return "", fmt.Errorf("internal error: did not receive auth token from server, please contact Modal support")
	}

	var expiry int64
	if exp := m.decodeJWT(token); exp > 0 {
		expiry = exp
	} else {
		m.logger.WarnContext(ctx, "x-modal-auth-token does not contain exp field")
		// We'll use the token, and set the expiry to 20 min from now.
		expiry = time.Now().Unix() + DefaultExpiryOffset
	}

	m.tokenAndExpiry.Store(TokenAndExpiry{
		token:  token,
		expiry: expiry,
	})

	timeUntilRefresh := time.Duration(expiry-time.Now().Unix()-RefreshWindow) * time.Second
	m.logger.DebugContext(ctx, "Fetched auth token",
		"expires_in", time.Until(time.Unix(expiry, 0)),
		"refresh_in", timeUntilRefresh)

	return token, nil
}

// Extracts the exp claim from a JWT token.
func (m *AuthTokenManager) decodeJWT(token string) int64 {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return 0
	}

	payload := parts[1]
	for len(payload)%4 != 0 {
		payload += "="
	}

	decoded, err := base64.URLEncoding.DecodeString(payload)
	if err != nil {
		return 0
	}

	var claims map[string]interface{}
	if err := json.Unmarshal(decoded, &claims); err != nil {
		return 0
	}

	if exp, ok := claims["exp"].(float64); ok {
		return int64(exp)
	}

	return 0
}

// GetCurrentToken returns the current cached token.
func (m *AuthTokenManager) GetCurrentToken() string {
	data := m.tokenAndExpiry.Load().(TokenAndExpiry)
	return data.token
}

// IsExpired checks if the current token is expired.
func (m *AuthTokenManager) IsExpired() bool {
	data := m.tokenAndExpiry.Load().(TokenAndExpiry)
	return isExpired(data)
}

func isExpired(data TokenAndExpiry) bool {
	return time.Now().Unix() >= data.expiry
}

func needsRefresh(data TokenAndExpiry) bool {
	return time.Now().Unix() >= data.expiry-RefreshWindow
}

// SetToken sets the token and expiry (for testing).
func (m *AuthTokenManager) SetToken(token string, expiry int64) {
	m.tokenAndExpiry.Store(TokenAndExpiry{
		token:  token,
		expiry: expiry,
	})
}

File: modal-js/src/auth_token_manager.ts
// Start refreshing this many seconds before the token expires
import type { Logger } from "./logger";

export const REFRESH_WINDOW = 5 * 60;
// If the token doesn't have an expiry field, default to current time plus this value (not expected).
export const DEFAULT_EXPIRY_OFFSET = 20 * 60;

export class AuthTokenManager {
  private client: any;
  private logger: Logger;
  private currentToken: string = "";
  private tokenExpiry: number = 0;
  private timeoutId: NodeJS.Timeout | null = null;
  private running: boolean = false;
  private fetchPromise: Promise<void> | null = null;

  constructor(client: any, logger: Logger) {
    this.client = client;
    this.logger = logger;
  }

  /**
   * Returns a valid auth token.
   * If the current token is expired and the manager is running, triggers an on-demand refresh.
   */
  async getToken(): Promise<string> {
    if (this.currentToken && !this.isExpired()) {
      return this.currentToken;
    }

    if (this.running) {
      await this.runFetch();
      if (this.currentToken && !this.isExpired()) {
        return this.currentToken;
      }
    }

    throw new Error("No valid auth token available");
  }

  /**
   * Fetches a new auth token from the server and stores it.
   */
  private async fetchToken(): Promise<void> {
    const response = await this.client.authTokenGet({});
    const token = response.token;

    if (!token) {
      throw new Error(
        "Internal error: did not receive auth token from server, please contact Modal support",
      );
    }

    this.currentToken = token;

    // Parse JWT expiry
    const exp = this.decodeJWT(token);
    if (exp > 0) {
      this.tokenExpiry = exp;
    } else {
      this.logger.warn("x-modal-auth-token does not contain exp field");
      // We'll use the token, and set the expiry to DEFAULT_EXPIRY_OFFSET from now.
      this.tokenExpiry = Math.floor(Date.now() / 1000) + DEFAULT_EXPIRY_OFFSET;
    }

    const now = Math.floor(Date.now() / 1000);
    const expiresIn = this.tokenExpiry - now;
    const refreshIn = this.tokenExpiry - now - REFRESH_WINDOW;
    this.logger.debug(
      "Fetched auth token",
      "expires_in",
      `${expiresIn}s`,
      "refresh_in",
      `${refreshIn}s`,
    );
  }

  /**
   * Background loop that refreshes tokens REFRESH_WINDOW seconds before they expire.
   */
  private async backgroundRefresh(): Promise<void> {
    while (this.running) {
      const now = Math.floor(Date.now() / 1000);
      const refreshTime = this.tokenExpiry - REFRESH_WINDOW;
      const delay = Math.max(0, refreshTime - now) * 1000;

      // Sleep until it's time to refresh
      await new Promise<void>((resolve) => {
        this.timeoutId = setTimeout(resolve, delay);
        this.timeoutId.unref();
      });

      if (!this.running) {
        return;
      }

      // Fetch new token
      try {
        await this.runFetch();
      } catch (error) {
        this.logger.error("Failed to refresh auth token", "error", error);
        // Sleep for 5 seconds before trying again on failure
        await new Promise((resolve) => setTimeout(resolve, 5000));
      }
    }
  }

  /**
   * Fetches the initial token and starts the refresh loop.
   * Throws an error if the initial token fetch fails.
   */
  async start(): Promise<void> {
    if (this.running) {
      return;
    }

    this.running = true;
    try {
      await this.runFetch();
    } catch (error) {
      this.running = false;
      throw error;
    }

    // Start background refresh loop, do not await
    this.backgroundRefresh();
  }

  /**
   * Stops the background refresh.
   */
  stop(): void {
    this.running = false;
    if (this.timeoutId) {
      clearTimeout(this.timeoutId);
      this.timeoutId = null;
    }
  }

  /**
   * Extracts the exp claim from a JWT token.
   */
  private decodeJWT(token: string): number {
    try {
      const parts = token.split(".");
      if (parts.length !== 3) {
        return 0;
      }

      let payload = parts[1];
      while (payload.length % 4 !== 0) {
        payload += "=";
      }

      const decoded = atob(payload);
      const claims = JSON.parse(decoded);
      return claims.exp || 0;
    } catch {
      return 0;
    }
  }

  isExpired(): boolean {
    const now = Math.floor(Date.now() / 1000);
    return now >= this.tokenExpiry;
  }

  private runFetch(): Promise<void> {
    if (!this.fetchPromise) {
      this.fetchPromise = (async () => {
        try {
          await this.fetchToken();
        } finally {
          this.fetchPromise = null;
        }
      })();
    }
    return this.fetchPromise;
  }

  getCurrentToken(): string {
    return this.currentToken;
  }

  setToken(token: string, expiry: number): void {
    this.currentToken = token;
    this.tokenExpiry = expiry;
  }
}
